src/Features/CSharp/Portable/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -1,214 +1,101 @@
-                            // the "&&" operator is applied. For example, in "if (expr && cond1 && cond2)" we need to
-                            // separate out the leftmost expression and reconstruct the rest of conditions into a new
-                            // expression, so that they would be evaluated entirely after "expr" e.g. "expr && (cond1 && cond2)".
-                            // Afterwards, we can directly use it in a case guard e.g. "case pat when cond1 && cond2:".
-                            var parentBinary = (BinaryExpressionSyntax)leftmost.WalkUpParentheses().Parent;
-                            var grandparentBinary = parentBinary.WalkUpParentheses().Parent as BinaryExpressionSyntax;
-
-                            var condition = grandparentBinary?.WithLeft(parentBinary.Right);
-
-                            return new Pattern.Guarded(pattern, (condition ?? node.Right).WalkDownParentheses());
-                        }
-
-                    default:
-                        return null;
-                }
+                return ReturnStatement(
+                    SwitchExpression((ExpressionSyntax)target, SeparatedList(sections.Select(AsSwitchExpressionArmSyntax))));
             }
 
-            // We do not offer a fix if the if-statement contains a break-statement, e.g.
-            //
-            //      while (...)
-            //      {
-            //          if (...) {
-            //              break;
-            //          }
-            //      }
-            // 
-            // When the 'break' moves into the switch, it will have different flow control impact.
-            protected override bool CanConvertIfToSwitch(IfStatementSyntax ifStatement)
-                => !_semanticModel.AnalyzeControlFlow(ifStatement).ExitPoints.Any(n => n.IsKind(SyntaxKind.BreakStatement));
-
-            protected override IEnumerable<ExpressionSyntax> GetLogicalOrOperands(
-                ExpressionSyntax syntaxNode)
+            private static SwitchExpressionArmSyntax AsSwitchExpressionArmSyntax(SwitchSection section)
             {
-                syntaxNode = syntaxNode.WalkDownParentheses();
-                while (syntaxNode.IsKind(SyntaxKind.LogicalOrExpression))
-                {
-                    var binaryExpression = (BinaryExpressionSyntax)syntaxNode;
-                    yield return binaryExpression.Right.WalkDownParentheses();
-                    syntaxNode = binaryExpression.Left.WalkDownParentheses();
-                }
-
-                yield return syntaxNode;
+                return SwitchExpressionArm(
+                    pattern: section.Labels.IsDefault
+                        ? DiscardPattern()
+                        : AsPatternSyntax(section.Labels[0].Pattern),
+                    whenClause: section.Labels.IsDefault
+                        ? null 
+                        : AsWhenClause(section.Labels[0]),
+                    expression: AsExpressionSyntax(section.Body));
```

Comment:
consider putting in a before/after comment here showing what this is generally doing.